### ISSUE TITLE
Add support for views to meta()

### ISSUE THREAD

**Issue by rbuehlma** [jOOQ/jOOQ#8404]
### Use case:

We are working on a migration tool which is able to do automated DDL migration to a predefined schema. To support multiple databases, we rely on jOOQ to generate the actual statements. jOOQ directly connects to the DB and is not used to generate any classes.

### Possible solution you'd like to see:

Add "getViews()" to meta() which allows to fetch all views including their definition. Since the definition of the view might contain proprietary SQL, it might not be possible to parse the definition but simply return it as string.

Additionally, it should be possible to create views using an sql string. Currently it only seems to be possible if the statement can be parsed by jOOQ (createView().as only accepts a Select<>).

Would be great if view dependencies can also be provided.

**Comment by lukaseder** [comment]
Thank you very much for your request. There are several requests here, which we'll definitely look into through separate tasks:

> Add "getViews()" to meta() which allows to fetch all views ...

These are two things. First off, you'd like to be able to distinguish between tables and views in jOOQ's meta model. There are different existing feature requests (e.g. #6412), which would like to see similar things. This is quite an impactful change and needs further thought.

Then:

> ...  including their definition. Since the definition of the view might contain proprietary SQL, it might not be possible to parse the definition but simply return it as string.

This would indeed be a very interesting addition once we have a view meta model that can be distinguished from the existing table meta model. I'm not quite sure whether to put this in a separate issue or do this with the various existing view / table distinction tasks.

> Additionally, it should be possible to create views using an sql string. Currently it only seems to be possible if the statement can be parsed by jOOQ (createView().as only accepts a Select<>).

Completely agree. Will implement through a separate issue: #8409 "Add support for plain SQL CREATE VIEW statements"

> Would be great if view dependencies can also be provided.

Would you mind elaborating on this?

**Comment by rbuehlma** [comment]
> > Would be great if view dependencies can also be provided.

> Would you mind elaborating on this?

At least Postgres provides a possibility the get the "base" relations (tables or views) on which a view relies. Would be great, if that information can be retrieved through Jooq. This is relevant when it comes to deciding which view needs to be created/dropped first.

**Comment by lukaseder** [comment]
So, in other words, you'd like to know for each column in each view what the transitive dependencies of that column are, e.g.

```sql
CREATE VIEW v1 AS SELECT c1 FROM t1
```

So, `v1.c1` depends on `t1.c1` and...

```sql
CREATE VIEW v2 AS SELECT v1.c1 + t2.c2 AS x FROM v1, t2
```

... `v2.x` depends on `v1.c1` (which depends on `t1.c1) and on `t2.c2`.

Is that it?

**Comment by rbuehlma** [comment]
From what I currently know,  the information is not required on field level but only per relation. in your sample,
`v1` depends on `t1` and
`v2` depends on `v1` and `t2`

Transitional dependencies (`v2` depends on `t1`) can then be resolved by the migration tool.

**Comment by lukaseder** [comment]
I see. Well, I think I'd prefer doing this completely with jOOQ, with the entire information available. If you only need tables, that's fine, that would be easy to offer through the jOOQ API from the complete information.

**Comment by lukaseder** [comment]
... obviously, this all applies to routines as well. A stored procedure has dependencies on other stored procedures, views, tables, etc. And views can have dependencies on stored functions. We might also want to include triggers, which introduce a dependency between tables (no views involved).

**Comment by rbuehlma** [comment]
Yes, you are right, there are also tables which depend on UDTs but this information is already in the jooq schema as soon as the UDT type is available.
In my opinion it would be good to have the possibility the get the direct and indirect dependencies separately.
Also stored procedures and triggers are not yet supported by the migration and would therefore be second priority for me.

**Comment by lukaseder** [comment]
Sure, I understand you have your priorities :) But such a significant feature needs to be implemented in the context of the big picture. I'll think about this and create a separate feature request with a bunch of bullets / TODOs, soon.

**Comment by lukaseder** [comment]
After a second review of this feature request, I think the requirements are clear and will be implemented through a variety of other issues:

- Clearly, we eventually need to be able to distinguish between tables and views (and others) more formally. That is already covered by #6412, #2370, and a few others. A variety of features can be derived from such a distinction, including
- The ability of managing a view's source code in jOOQ, either as raw text or in a parsed version that uses the jOOQ API in generated code
- Once we have a jOOQ API version (parsed implicitly by the code generator or explicitly by yourself), we will be able to traverse the view text. Right now, this can be done already with a `VisitListener`, which you can use on your side, if you want. In the future, the internal expression tree model will be improved such that a `VisitListener` is no longer necessary and a simpler API can be used to find tables in a view.

Everything needed in this requirement is on the roadmap in one form or the other, so I'm closing this as a duplicate of other issues.

**Comment by lukaseder** [comment]
#8409 (Add support for plain SQL CREATE VIEW statements) is implemented in jOOQ 3.12
